lib/date_time/date_time.c

```diff
@@ -64,8 +80,12 @@ static struct sntp_time sntp_time;
 K_SEM_DEFINE(time_fetch_sem, 0, 1);
 
 static struct k_work_delayable time_work;
+#if defined(CONFIG_DATE_TIME_AUTO)
+static struct k_work init_auto_work;
+#endif
 
 static struct time_aux {
```

Comment:
Zephyr time or modem time as source if time_aux is removed
---------------
From TommiK: One more thing which I think we should consider is whether modem would be more accurate in keeping track of the time. This is possible, I think it's even probable. I can do a test run during the weekend by disabling network time updates from the modem and then set the same time to both modem and application side and see what happens in a day or two.

Reply:
Is it reasonable to depend on the modem keeping track of time? What happens in the scenario where the application wants to shut down the modem for some reason, would that work?

Reply:
Fair points, @trantanen and @simensrostad. It's probably not a good idea to issue an AT command whenever `date_time_now()` is called. Modem seems to return time when it's in offline mode (AT+CFUN=4), but not when it's powered off (AT+CFUN=0). I think these are both good reasons to keep track of the time on the application side.

However, based on a test run during the weekend, the modem clock is much more accurate than the application side clock. I did a test run which lasted about 64 hours. During that time the modem time drifted forward ~1 second, but the application clock ~12 seconds. For the test I disabled network time and enabled PSM, so that modem would be spending most of the time in deep sleep. Based on this, it could make sense to synchronize the application clock from the modem clock periodically. Or we could just listen to the XTIME notifications and synchronize the application clock from that each time. This still leaves the case where time needs to be obtained using NTP. In that case accuracy could be improved by setting the modem clock and then synchronizing time back to the application side from the modem periodically.

Reply:
@trantanen I'm still concerned by one thing. When modem doesn't get time from the network, the library gets time from NTP and sets also the modem clock. This changes the behavior of the periodic synchronization, because after modem time has been set once, the library would always use time from the modem for the periodic synchronization. On the long run, this will cause the clock to drift, because device is not getting updates from the network nor NTP.

This could be mitigated by using the XTIME notification to determine whether modem gets time updates from the network. If at least one XTIME notification has been received, periodic synchronization would use AT+CCLK? like it does now. However, if no XTIME notifications have been received, the updates would use NTP (and also update the time to the modem). What do you think?

Reply:
This is still a bit tricky or requires kind of hack to deal with. We have information when modem time is valid and that's purely based on XTIME notification. However, when periodic update occurs, we first check modem and then NTP. If modem succeeds, we use that no matter whether modem got the time from network or whether we set it based on NTP time. If we know modem time is set based on NTP time, we could go for NTP but what if that fails. Then, we'd like to use the modem time set based on NTP. This is a bit of a code structure issue meaning it'll likely get a bit dirty to fix this.

Reply:
The resolution here is that modem time is not used before we get XTIME notification.
This discussion is ready from my side. @tokangas Please close or comment further.